services/src/main/java/org/keycloak/authentication/requiredactions/UpdateUserLocaleAction.java

```diff
@@ -0,0 +1,66 @@
+package org.keycloak.authentication.requiredactions;
+
+import org.keycloak.Config;
+import org.keycloak.authentication.RequiredActionContext;
+import org.keycloak.authentication.RequiredActionFactory;
+import org.keycloak.authentication.RequiredActionProvider;
+import org.keycloak.locale.LocaleSelectorProvider;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.models.UserModel;
+
+public class UpdateUserLocaleAction implements RequiredActionProvider, RequiredActionFactory {
+
+    @Override
+    public String getDisplayText() {
+        return "Update User Locale";
+    }
+
+    @Override
+    public void evaluateTriggers(RequiredActionContext context) {
+        String userRequestedLocale = context.getAuthenticationSession().getAuthNote(LocaleSelectorProvider.USER_REQUEST_LOCALE);
```

Comment:
RequiredAction is basically used here as a callback to be triggered after authentication. This works fine, on the other hand it looks to me as a workaround. It has some unneeded side-effects (EG. administrator will see "update user locale" as requiredAction and they will be able to set it to users. This won't have any effect and might be slightly confusing etc).

It may be better to have something like "Interceptor SPI" or possibly re-use our EventSPI. However using Event SPI in it's current form looks also as a workaround an I guess we don't have proper event yet for "POST_REQUIRED_ACTIONS", which is triggered after success of all required actions etc.

In shortcut, consider that we don't have better SPI right now, I am fine with using it as it for this PR :)

Reply:
What I want to do is make required action API into action API. Where an action can advertise how it supports being triggered (i.e. an action after auth, an required action, an aia, etc.). Think that is better than yet another API

Reply:
Also, someone may want to remove this action and replace it with something else that prompts the user with a form.